babel-plugin-transform-jsx/lib/index.js

```diff
@@ -154,21 +58,139 @@ export default function ({ types: t }) {
   }
 
   /* ==========================================================================
+   * Visitors
+   * ======================================================================= */
+
+  const visitJSXElement = (path, state) => {
+
+    const {
+      opts: {
+        useVariables = false
+      }
+    } = state
+
+    /* ==========================================================================
+     * Node Transformers
+     * ======================================================================= */
+
+    const JSXIdentifier = node => t.stringLiteral(node.name)
+
+    const JSXNamespacedName = node => t.stringLiteral(`${node.namespace.name}:${node.name.name}`)
+
+    const JSXMemberExpression = transformOnType({
+      JSXIdentifier: node => t.identifier(node.name),
+      JSXMemberExpression: node => (
+        t.memberExpression(
+          JSXMemberExpression(node.object),
+          JSXMemberExpression(node.property)
+        )
+      )
+    })
+
+    const JSXElementName = transformOnType({
+      JSXIdentifier: useVariables
+        ? node => path.scope.hasBinding(node.name) ? t.identifier(node.name) : JSXIdentifier(node)
```

Comment:
TLDR: I now prefer your solution and propose option names and default at the end.

It was just the way I naturally thought of the feature.
1. Have component references
2. Want to reference them as elements
3. Need to differentiate between components and normal elements
4. Two ideas
  1. List normal elements and check for inclusion in the list -- would limit users to recognized list or require them to maintain their own list
  2. Look for variables in scope
5. Chose the latter as more maintainable

There was a little more to it, but that was my basic reasoning. After hearing your idea, I'm not sure which I prefer. I'll list some pros and cons for each.

# Check for bound names

* *Pros*
  * Requires no further configuration, just uses bound names if they're in scope
  * Based firmly in the reality of the scope (except global), rather than external naming rules
    * Is this ultimately more intuitive?
* *Cons*
  * Variables named after HTML elements will mask use of the element names
    * Would beg for a lint rule to warn about these cases
  * Denies the ability to reference global components
    * We could argue this is good, but I think it is bad to remove the choice.


# Use names that match a pattern

* *Pros*
  * Configuration makes an intentional statement about how components should be named
  * Can naturally avoid overlap with HTML element names by choosing a pattern that does not match HTML element names. Examples:
    * First letter uppercase
    * First character an ampersand (which I didn't know was permissable, but I'm trusting you :) )
  * Would not artificially deny use of global components
  * React appears to rely on naming convention in my tests. Lowercase names were not recognized as components, regardless of whether the lowercase names were valid HTML element names.
* *Cons*
  * Not based in the reality of the scope, can reference undeclared names
    * Would beg for a lint rule to warn about these cases

There are probably more, but this is enough for me.

# Thoughts

I prefer your solution because it allows clear separation of component and normal elements in the element namespace. In addition, I like the idea of uppercase indicating components and lowercase indicating normal elements and propose that as the default. What do you think?

If we go this route, I think we should have two options, one to enable the feature and the other to specify a custom naming pattern.

Following the proposal in my separate comment about option naming, what do you think of the following?
* allowComponentElements
* componentElementNamePattern

The latter feels a too wordy to me, but I like its specificity.